presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java

```diff
@@ -176,7 +178,13 @@ private boolean arePlansSame(FilterNode filter, TableScanNode tableScan, PlanNod
                             // Simplify the tuple domain to avoid creating an expression with too many nodes,
                             // which would be expensive to evaluate in the call to isCandidate below.
                             domainTranslator.toPredicate(newDomain.simplify().transform(assignments::get))));
-            constraint = new Constraint(newDomain, evaluator::isCandidate);
+            Optional<DiscretePredicates> discretePredicates = metadata.getTableProperties(session, node.getTable()).getDiscretePredicates();
+            if (discretePredicates.isPresent() && Collections.disjoint(discretePredicates.get().getColumns(), evaluator.getArguments())) {
+                constraint = new Constraint(newDomain);
+            }
+            else {
+                constraint = new Constraint(newDomain, evaluator::isCandidate);
+            }
```

Comment:
Why is this needed? If possible, we want to avoid using the `getDiscretePredicates()` API since we're trying to get rid of it.

Reply:
Something like this is needed to prevent a predicate from being created on the Constraint and causing the partition filter check from incorrectly passing. See [these](https://github.com/prestosql/presto/pull/2334/files#diff-f735a7bf0d40c13be5c34d9a25110a08R366-R414) tests for examples. The first test is an example of a query that should pass the check (since it's pruning partitions) while the second shouldn't pass (since it's not even filtering on a partition column). The current logic in Presto creates a predicate that always returns true in the case of the filter not containing a reference to any partition columns (for Hive) [here](https://github.com/prestosql/presto/pull/2334/files#diff-64a95c3a25c48aeecc5f744e3e45b7b8L298-L300). The check I implemented [here](https://github.com/prestosql/presto/pull/2334/files#diff-7e85fbe1d8919e64b7cf7bcb615e12f4R197) differentiates these based on the existence of a predicate on the constraint.

Maybe this could be a method on ConnectorMetadata instead?

Reply:
Yeah, this is problematic because it might break connectors that can leverage the LayoutEvaluator::isCandidate logic but do not expose DiscretePredicates (which are a hack that only exists for legacy reasons).

Reply:
I think what we could do is extend Constraint to include the list of columns included in the filter expression. With that, the connector should be able to keep track of which columns are being considered for filtering and fail the query during the call to `validateScan` if at least one partition column doesn't appear.